dipy/tracking/utils.py

```diff
@@ -160,37 +165,87 @@ def connectivity_matrix(streamlines, affine, label_volume,
-
-    # get labels for label_volume
-    i, j, k = endpoints.T
-    endlabels = label_volume[i, j, k]
-    if symmetric:
-        endlabels.sort(0)
-    mx = label_volume.max() + 1
-    matrix = ndbincount(endlabels, shape=(mx, mx))
-    if symmetric:
-        matrix = np.maximum(matrix, matrix.T)
-
-    if return_mapping:
-        mapping = defaultdict(list)
-        for i, (a, b) in enumerate(endlabels.T):
-            mapping[a, b].append(i)
-
-        # Replace each list of indices with the streamlines they index
-        if mapping_as_streamlines:
-            for key in mapping:
-                mapping[key] = [streamlines[i] for i in mapping[key]]
-
-        # Return the mapping matrix and the mapping
-        return matrix, mapping
+    if inclusive:
+        # Create ndarray to store streamline connections
+        edges = np.ndarray(shape=(3, 0), dtype=int)
+        lin_T, offset = _mapping_to_voxel(affine)
+        for sl in range(len(streamlines)):
+            # Convert streamline to voxel coordinates
+            entire = _to_voxel_coordinates(streamlines[sl], lin_T, offset)
+            i, j, k = entire.T
+
+            if symmetric:
+                # Create list of all labels streamline passes through
+                entirelabels = list(OrderedDict.fromkeys(label_volume[i, j, k]))
+                # Append all connection combinations with streamline number
+                for comb in combinations(entirelabels, 2):
+                    edges = np.append(edges, [[comb[0]], [comb[1]], [sl]],
+                                      axis=1)
+            else:
+                # Create list of all labels streamline passes through, keeping
+                # order and whether a label was entered multiple times
+                entirelabels = list(groupby(label_volume[i, j, k]))
+                # Append connection combinations along with streamline number,
+                # removing duplicates and connections from a label to itself
+                combs = set(combinations([z[0] for z in entirelabels], 2))
+                for comb in combs:
+                    if comb[0] == comb[1]:
+                        pass
+                    else:
+                        edges = np.append(edges, [[comb[0]], [comb[1]], [sl]],
+                                          axis=1)
+        if symmetric:
+            edges[0:2].sort(0)
+        mx = label_volume.max() + 1
+        matrix = ndbincount(edges[0:2], shape=(mx, mx))
+
+        if symmetric:
+            matrix = np.maximum(matrix, matrix.T)
+        if return_mapping:
+            mapping = defaultdict(list)
+            for i, (a, b, c) in enumerate(edges.T):
+                mapping[a, b].append(c)
+            # Replace each list of indices with the streamlines they index
+            if mapping_as_streamlines:
+                for key in mapping:
+                    mapping[key] = [streamlines[i] for i in mapping[key]]
+
+            return matrix, mapping
+        else:
```

Comment:
this last `else` is not needed. You can just return

Reply:
Done. https://github.com/nipy/dipy/pull/2012/commits/369eb44d5b416175dbccebf6f3c88f1d6ba25084